File: algorithm/337.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <ctype.h>
#include <math.h>
struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

#define MAX(x, y) (x < y ? y : x)

int rob(struct TreeNode *root)
{
    if (root == NULL)
    {
        return 0;
    }
    int res1 = root->val;
    if (root->left != NULL)
    {
        res1 += (rob(root->left->left) + rob(root->left->right));
    }
    if (root->right != NULL)
    {
        res1 += (rob(root->right->left) + rob(root->right->right));
    }

    int res2 = rob(root->left) + rob(root->right);
    return MAX(res1, res2);
}

File: go/goalgor/algor/337.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func Max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
func rob(root *TreeNode) int {
	var dfs func(*TreeNode) []int
	dfs = func(root *TreeNode) []int {
		if root == nil {
			return []int{0, 0}
		}
		l := dfs(root.Left)
		r := dfs(root.Right)
		yes := root.Val + l[1] + r[1]
		no := Max(l[0], l[1]) + Max(r[0], r[1])
		return []int{yes, no}
	}
	ans := dfs(root)
	return Max(ans[0], ans[1])
}